File: ICPC/WarmUp 10/ncpc2015solutions/bells_lukas.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#include <cstdio>
using namespace std;

#define rep(i,a,b) for(auto i=a; i<(b); ++i)

typedef vector<int> vi;
vector<vi> gen(int n)
{
    if (n == 1) return vector<vi> (1, vi(1, 1));
    vector<vi> next = gen(n - 1);

    vector<vi> res;
    bool dir = false;
    for (vi &o : next)
    {
        if (!dir) o.push_back(n);
        else o.insert(o.begin(), n);
        res.push_back(o);

        int cur = dir ? 0 : n - 1;
        rep(i,0,n-1)
        {
            int pos = cur + (dir ? 1 : -1);
            swap(o[cur], o[pos]);
            cur = pos;
            res.push_back(o);
        }
        dir = !dir;
    }
    return res;
}
int main()
{
    int n;
    scanf("%d", &n);

    vector<vi> res = gen(n);
    for (vi &o : res)
    {
        rep(i,0,n-1) cout << o[i] << " ";
        cout << o.back() << endl;
    }
}

File: ICPC/WarmUp 10/ncpc2015solutions/bells_Lukas.java
import java.util.*;
import java.util.stream.*;
import java.io.*;

public class Lukas
{
  public static List<List<Integer>> gen(int n) {
    List<List<Integer>> res = new ArrayList<List<Integer>> ();
    if (n == 1) {
      List<Integer> tmp = new ArrayList<Integer> ();
      tmp.add(1);
      res.add(tmp);
      return res;
    }

    List<List<Integer>> next = gen(n - 1);
    boolean dir = false;
    for (List<Integer> o : next) {
      if (!dir) o.add(n);
      else o.add(0, n);
      res.add(o);

      int cur = dir ? 0 : n - 1;
      for (int i = 0; i < n - 1; i++) {
        int pos = cur + (dir ? 1 : -1);
        List<Integer> newo = new ArrayList<Integer> (o);

        int tmp = newo.get(cur);
        newo.set(cur, newo.get(pos));
        newo.set(pos, tmp);

        cur = pos;
        res.add(newo);
        o = newo;
      }
      dir = !dir;
    }
    return res;
  }

  public static void main(String[] args) {
    final int n = (new Scanner(System.in)).nextInt();
    PrintWriter out = new PrintWriter(new BufferedOutputStream(System.out));

    List<List<Integer>> res = gen(n);
    for (List<Integer> o : res) {
      for (int x : o)
        out.print(x + " ");
      out.println();
    }

    out.flush();
  }
}
